internal/keepers/service.go

```diff
@@ -53,38 +68,27 @@ func newOnDemandUpkeepService(ratio sampleRatio, registry types.Registry, logger
 
 	// start background services
 	s.start()
+
 	return s
 }
 
 var _ upkeepService = (*onDemandUpkeepService)(nil)
 
 func (s *onDemandUpkeepService) SampleUpkeeps(ctx context.Context, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	// get only the active upkeeps from the contract. this should not include
-	// any cancelled upkeeps
-	keys, err := s.registry.GetActiveUpkeepKeys(ctx, "0")
-	if err != nil {
-		return nil, fmt.Errorf("%w: failed to get upkeeps from registry for sampling", err)
-	}
-
-	s.logger.Printf("%d active upkeep keys found in registry", len(keys))
-	if len(keys) == 0 {
-		return []*types.UpkeepResult{}, nil
+	if s.workers == nil {
+		panic("cannot sample upkeeps without runner")
 	}
 
-	// select x upkeeps at random from set
-	keys = s.shuffler.Shuffle(keys)
-	size := s.ratio.OfInt(len(keys))
-
-	s.logger.Printf("%d keys selected by provided ratio %s", size, s.ratio)
-	if size <= 0 {
-		return []*types.UpkeepResult{}, nil
-	}
+	s.eligibleUpkeepKeysLock.Lock()
+	keys := make([]types.UpkeepKey, len(s.eligibleUpkeepKeys))
+	copy(keys, s.eligibleUpkeepKeys)
+	s.eligibleUpkeepKeysLock.Unlock()
```

Comment:
Instead of using properties of the service, create a data structure that has the single job of maintaining and producing eligible keys. It should have its own internal locks, but should also purge keys like the cache.

Reply:
Done